ScoutSuite/__main__.py

```diff
@@ -0,0 +1,210 @@
+            cloud_provider.services = last_run_dict['services']
+            for service in cloud_provider.service_list:
+                cloud_provider.services[service] = current_run_services[service]
+
+    # Partial run, using pre-pulled data
+    else:
+        print_info('Using local data')
+        # Reload to flatten everything into a python dictionary
+        last_run_dict = report.encoder.load_from_file(ReportFile.results)
+        for key in last_run_dict:
+            setattr(cloud_provider, key, last_run_dict[key])
+
+    # Pre processing
+    cloud_provider.preprocessing(
+        args.get('ip_ranges'), args.get('ip_ranges_name_key'))
+
+    # Analyze config
+    print_info('Running rule engine')
+    finding_rules = Ruleset(environment_name=args.get('profile'),
+                            cloud_provider=args.get('provider'),
+                            filename=args.get('ruleset'),
+                            ip_ranges=args.get('ip_ranges'),
+                            aws_account_id=cloud_provider.aws_account_id)
+    processing_engine = ProcessingEngine(finding_rules)
+    processing_engine.run(cloud_provider)
+
+    # Create display filters
+    print_info('Applying display filters')
+    filter_rules = Ruleset(cloud_provider=args.get('provider'),
+                           filename='filters.json',
+                           rule_type='filters',
+                           aws_account_id=cloud_provider.aws_account_id)
+    processing_engine = ProcessingEngine(filter_rules)
+    processing_engine.run(cloud_provider)
+
+    if args.get('exceptions')[0]:
+        print_info('Applying exceptions')
+        try:
+            exceptions = RuleExceptions(
+                args.get('profile'), args.get('exceptions')[0])
+            exceptions.process(cloud_provider)
+            exceptions = exceptions.exceptions
+        except Exception as e:
+            print_debug(
+                'Failed to load exceptions. The file may not exist or may have an invalid format.')
+            exceptions = {}
+    else:
+        exceptions = {}
+    # Handle exceptions
+    try:
+        exceptions = RuleExceptions(
+            args.get('profile'), args.get('exceptions')[0])
+        exceptions.process(cloud_provider)
+        exceptions = exceptions.exceptions
+    except Exception as e:
+        print_debug(
+            'Warning, failed to load exceptions. The file may not exist or may have an invalid format.')
+        exceptions = {}
+
+    # Finalize
+    cloud_provider.postprocessing(report.current_time, finding_rules)
+
+    # Save config and create HTML report
+    html_report_path = report.save(cloud_provider, exceptions, args.get('force_write'),
+                                   args.get('debug'))
+
+    # Open the report by default
+    if not args.get('no_browser'):
+        print_info('Opening the HTML report')
+        url = 'file://%s' % os.path.abspath(html_report_path)
+        webbrowser.open(url, new=2)
```

Comment:
Seems like this does not work if the server is not running. Couldn't we start the server automatically at this point, or add an argument that does that? Having to use two terminals and two commands is a bit of a pain.

Reply:
Oh right 2 terminals, sorry I wrote that because it's easier when developing, but the user can use the same terminate to generate the report and then --serve and open the report to view it. When talking with @j4v we chose to keep this in 2 steps if irc.